gadget/mountedfilesystem.go

```diff
@@ -186,3 +205,487 @@ func (m *MountedFilesystemWriter) writeVolumeContent(whereDir string, content *V
+		return f.backupOrCheckpointPrefix(dstRoot, target, backupDir)
+	}
+
+	// backup of the original content
+	backup, err := makeStampFile(backupName)
+	if err != nil {
+		return fmt.Errorf("cannot create backup file: %v", err)
+	}
+	// becomes a noop if canceled
+	defer backup.Commit()
+
+	// checksum the original data while it's being copied
+	origHash := crypto.SHA1.New()
+	htr := io.TeeReader(orig, origHash)
+
+	_, err = io.Copy(backup, htr)
+	if err != nil {
+		backup.Cancel()
+		return fmt.Errorf("cannot backup original file: %v", err)
+	}
+
+	// digest of the update
+	updateDigest, _, err := osutil.FileDigest(srcName, crypto.SHA1)
+	if err != nil {
+		backup.Cancel()
+		return fmt.Errorf("cannot checksum update file: %v", err)
+	}
+	// digest of the currently present data
+	origDigest := origHash.Sum(nil)
+
+	if bytes.Equal(origDigest, updateDigest) {
+		// files are identical, no backup needed
+		if err := makeStamp(sameStamp); err != nil {
+			return fmt.Errorf("cannot create a checkpoint file: %v", err)
+		}
+
+		// makes the previous commit a noop
+		backup.Cancel()
+		return nil
+	}
+
+	return f.backupOrCheckpointPrefix(dstRoot, target, backupDir)
+}
+
+func (f *MountedFilesystemUpdater) backupOrCheckpointVolumeContent(whereDir string, content *VolumeContent, preserveInDst []string, backupDir string) error {
+	if err := checkContent(content); err != nil {
+		return err
+	}
+
+	srcName := f.entrySource(content.Source)
+
+	if osutil.IsDirectory(srcName) || strings.HasSuffix(content.Source, "/") {
+		// backup directory contents
+		return f.backupOrCheckpointDirectory(whereDir, content.Source, content.Target, preserveInDst, backupDir)
+	} else {
+		// deploy a file
+		return f.backupOrCheckpointFile(whereDir, content.Source, content.Target, preserveInDst, backupDir)
+	}
+}
+
+// Rollback attempts to revert changes done by the update step, using state
+// information collected during backup phase. Files that were modified by the
+// update are stored from their backup copies, newly added directories are
+// removed.
+func (f *MountedFilesystemUpdater) Rollback() error {
+	mount, err := f.mountLookup(f.ps)
+	if err != nil {
+		return fmt.Errorf("cannot find mount location of structure %v: %v", f.ps, err)
+	}
+
+	backupRoot := fsStructBackupPath(f.backupDir, f.ps)
+
+	preserveInDst := prefixPreserve(mount, f.ps.Update.Preserve)
+
+	for _, c := range f.ps.Content {
+		if err := f.rollbackVolumeContent(mount, &c, preserveInDst, backupRoot); err != nil {
+			return fmt.Errorf("cannot rollback content: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (f *MountedFilesystemUpdater) rollbackPrefix(dstRoot, target string, backupDir string) error {
+	for prefix := filepath.Dir(target); prefix != "/" && prefix != "."; prefix = filepath.Dir(prefix) {
+		_, prefixDst, prefixBackupBase := f.entryNames(dstRoot, "", prefix, backupDir)
+		if !osutil.FileExists(prefixBackupBase + ".backup") {
+			// try remove
+			if err := os.Remove(prefixDst); err != nil {
```

Comment:
so newly created directories will end up without a .backup file and removed here?

Reply:
Yes, basically cleaning up newly created parts of the tree in the destination location.